x-pack/plugins/fleet/server/routes/epm/handlers.ts

```diff
@@ -451,8 +456,45 @@ export const installPackageByUploadHandler: FleetRequestHandler<
   const archiveBuffer = Buffer.from(request.body);
   const spaceId = fleetContext.spaceId;
   const user = (await appContextService.getSecurity()?.authc.getCurrentUser(request)) || undefined;
+  const logger = appContextService.getLogger();
 
   const authorizationHeader = HTTPAuthorizationHeader.parseFromRequest(request, user?.username);
+  const timeToWaitString = moment
+    .utc(moment.duration(UPLOAD_RETRY_AFTER_MS).asMilliseconds())
+    .format('mm[m]ss[s]');
+
+  // Find the latest packages installed by upload
+  const latestUploadedPackage = await savedObjectsClient.find<Installation>({
```

Comment:
this check prevents a second installation if the SO is already written out, it doesn't prevent if the upload starts twice quickly before the SO is created, isn't it?

Reply:
Yes I think that the first time when the SO is not written out yet the upload could be allowed more than once. This could happen only the first time ever that an install by upload is performed, as the query gets all the SO where `install_source = upload`

Reply:
+1 on @juliaElastic point here I think what we really want to prevent is to have concurrent install of upload package on the same Kibana instance, I think it could probably be simpler/better to implement something we global variable here instead of relying on the installation saved object

Reply:
I updated the PR, instead of reading from the SO it's now using an in memory cache that saves the timestamp of the latest install by upload.